Issue #123: Add `anib`

baileythegreen opened the pull request:
Adds `anib` subcommand to v3.

Closes #123.

## Type of change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality not to work as expected)
- [x] This change requires a documentation update
- [ ] This is a documentation update

## Action Checklist

- [x] Work on a single issue/concept (if there are multiple separate issues to address, please use a separate pull request for each)
- [ ] Fork the `pyani` repository under your own account (please [allow write access for repository maintainers](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork))
- [x] Set up an appropriate development environment (please see `CONTRIBUTING.md`)
- [x] Create a new branch with a short, descriptive name
- [x] Work on this branch
  - [x] style guidelines have been followed
  - [x] new code is commented, especially in hard-to-understand areas
  - [x] corresponding changes to documentation have been made
  - [x] tests for the change have been added that demonstrate the fix or feature works
- [x] Test locally with `pytest -v` **non-passing code will not be merged**
- [x] Rebase against `origin/master`
- [x] Check changes with `flake8` and `black` before submission
- [x] Commit branch
- [ ] Submit pull request, describing the content and intent of the pull request
- [ ] Request a code review
- [ ] Continue the discussion at [`Pull requests` section](https://github.com/widdowquinn/pyani/pulls) in the `pyani` repository

baileythegreen committed:
Initial split of v2 ANIb code into `anib.py` and `aniblastall.py` files

Including changing the parameter `mode` to `method`

baileythegreen committed:
Updated comments/logging

baileythegreen committed:
Added `sysexit` when adding genomes to the database fails

baileythegreen committed:
Added initial versions of functions and comments about needed code

`run_anib_jobs()`, `update_comparison_results()`

baileythegreen committed:
Rename `fraglens` to `fragsizes` for consistency with `anib.py`

baileythegreen committed:
Add `add_blastdb()` to `pyani_orm.py`

baileythegreen committed:
Change default for `maxmatch` from `None` to `False`

If this is not a boolean, it prevents sqlite from recognising duplicate 
entries

baileythegreen committed:
Add/expand code to process input genomes

Up to adding them to the database

baileythegreen committed:
Split genome files into contiguous fragments

baileythegreen committed:
Implement `generate_joblist()`

baileythegreen committed:
Implement `run_anib_jobs()`

baileythegreen committed:
Implement `update_comparisons_results()` and commit to database

baileythegreen committed:
Update call to `generate_joblist()`

Now passes the lists of all `fragfiles` and `fraglens`

baileythegreen committed:
Update name of output file in `fragment_fasta_file()`

baileythegreen committed:
Update value passed for `maxmatch` to a boolean

Necessary for sqlite to recognise duplicate values

baileythegreen committed:
Implement remainder of `subcmd_anib()`

baileythegreen committed:
Add a commented question

baileythegreen committed:
Alter `generate_blastn_commands()` to only take one query/subject pair

baileythegreen committed:
Alter `construct_blastn_cmdline()` for a single query/subject pair

baileythegreen committed:
Remove `method` parameter from `process_blast()` call

baileythegreen committed:
Change `outfilename` in `construct_makeblastdb_cmd()
`

baileythegreen committed:
Move `aniblastall`-specific tests to `test_aniblastall.py`

baileythegreen committed:
Change call to `parse_blast_tab()` to not use `method` parameter

baileythegreen committed:
Skip tests as a result of changes to how command lines are generated

Currently, we don't pass a list of genome pairs to `anib.py`.

baileythegreen committed:
Add `get_version()` tests and boilerplate

baileythegreen committed:
Merge branch 'master' of https://github.com/widdowquinn/pyani into anib_123

baileythegreen commented:
I am running into an issue:

- In the monolithic v2, `anib` and `aniblastall` were conjoined; we discussed separating them for v3.
- `pyani/scripts/average_nucleotide_identity.py` exists in the v3 branch, and both it, and its tests, expect `anib` and `aniblastall` to be conjoined.

As of yet, I am not sure how to separate the two, while keeping backwards-compatibility and also not repeating a lot of code.

baileythegreen committed:
Make changes to reflect splitting of `anib` and `aniblastall`

- remove `mode` or `method` arguments in function calls/definitions
- change references to module names from `anib` to `aniblastall` where 
necessary

baileythegreen commented:
I have resolved the issue with the legacy script in a way that I consider to be elegant; that works; and is fairly clear, but also only changes internals. However, it causes a minor problem with either `flake8` or CodeFactor. `flake8` doesn't seem to register the `__import__()` command as adding something to the namespace; and CodeFactor complains about the `global` statement I had to add to satisfy `flake8` so I could commit.

It passes all tests, though; and kept the changes to a minimum.

baileythegreen committed:
Add tests for `get_version()`

These got removed somewhere?

baileythegreen committed:
Fixed typos and put one variable name in camelcase

baileythegreen committed:
Fix which file is passed as input and which as blastdb

baileythegreen committed:
Move serialisation of fragment dictionary to primary function

baileythegreen committed:
Added useful debugging lines to `anib.py`

May be removed later on

baileythegreen committed:
Add long versions of CLI flags to `anib_parser.py`

baileythegreen committed:
Comment out (likely) unnecessary function

Will be removed later

baileythegreen committed:
Update docstrings to reflect new comparison handling

baileythegreen committed:
Add useful debugging lines

May be removed later

baileythegreen committed:
Populate `aniblastall_parser.py`

baileythegreen committed:
Comment out (likely) unnecessary function

Will be removed later

baileythegreen committed:
Simplify generation of command lines and update supporting docstrings

baileythegreen committed:
Update command name referenced in docstrings

baileythegreen committed:
Fix which files are passed as input and blastdb in command generation

baileythegreen committed:
Add useful debugging lines

May be removed later

baileythegreen committed:
Add imports and convenience struct for `ComparisonJob`

baileythegreen committed:
Add initial code for starting a run and adding it to the database

baileythegreen committed:
Add genomes for the ru  to the database

baileythegreen committed:
Get genomes for the run and create output directories

baileythegreen committed:
Add code to fragment fasta files

baileythegreen committed:
Add code to create `blastdb`

baileythegreen committed:
Add code to generate list of comparisons and filter existing ones

baileythegreen committed:
Add code for case where all comparisons have been run already

baileythegreen committed:
Add code for recovery mode

baileythegreen committed:
Rename `blastcmd` to `blastallcmd`

baileythegreen committed:
Add `generate_joblist()`

baileythegreen committed:
Add code to generate `joblist`

baileythegreen committed:
Add code to run comparisons

baileythegreen committed:
Add code to update database

baileythegreen committed:
Remove or rename things specific to `anib` to `aniblastall`

baileythegreen committed:
Skip unnecessary tests

Will probably be removed

baileythegreen committed:
Add `dir_aniblastall_in()` fixture

baileythegreen commented:
Reminder to myself that this is currently missing `test_aniblastall.py`, which probably explains (most of) the decrease in coverage.

baileythegreen committed:
Add lines debugging

May be removed later

baileythegreen committed:
Fix naming of elements in `test_subcmd_anib" namespace

baileythegreen committed:
Update naming of `TestANIbSubcommand()`

baileythegreen committed:
Switch to use `NamedTuple`

baileythegreen committed:
Add `test_subcmd_10_aniblastall.py`

baileythegreen committed:
Make `test_subcmd_04_anim.py` match other method subcommand tests

baileythegreen committed:
Fix names of executable variables

baileythegreen commented:
Failing test for `get_version()` on Linux because the information is in stdout, not stderr.

baileythegreen committed:
Pass both stdout and stderr to the regex version search

baileythegreen committed:
Set `indir` and `outdir` to required in `anib_parser.py`

baileythegreen committed:
Add `docs/subcmd_anib.rst`

baileythegreen committed:
Add documentation for `aniblastall`

baileythegreen committed:
Merge branch 'master' of https://github.com/widdowquinn/pyani into anib_123

baileythegreen committed:
Replace f-strings in logging statements

baileythegreen committed:
Update .gitignore

baileythegreen committed:
Merge branch 'master' of https://github.com/widdowquinn/pyani into anib_123

baileythegreen committed:
Update call to `add_run()` to fit the function's new return value

baileythegreen committed:
Remove f-strings from logging calls

baileythegreen committed:
Merge branch 'master' of https://github.com/widdowquinn/pyani into anib_123

baileythegreen committed:
Merge branch 'master' of https://github.com/widdowquinn/pyani into anib_123

baileythegreen commented:
> @widdowquinn I think this is ready to merge. The things being caught by codefactor are calls to `subprocess.run` that use `shell=True`, which I don't think we can do differently, and the use of `global anib` in `average_nucleotide_identity.py`, which is necessary to maintain backwards compatibility so the legacy tests pass.

This is once again the state of this PR. Can it be merged?

widdowquinn reviewed (changes requested):
If tests are no longer relevant or applicable, they should be removed or updated to make them relevant. The "unsure this is needed" message here ([cc4d346](https://github.com/widdowquinn/pyani/pull/338/commits/cc4d3465798e9df5612d1c2dc133893d5a72c1b3) and [3233fd5](https://github.com/widdowquinn/pyani/pull/338/commits/3233fd5c7db5f15389daa162434ac1468f2f4b6d)) reads as though the person making the change does not know what the test does, and/or why it fails or might be important.

In this case, the intent of the tests was to supplement the corresponding `*single()` function (which tests generation of a single command line, given input files), to check whether the appropriate reciprocal BLAST comparison command lines were being correctly generated. This allowed some level of diagnosis when a comparison fails to establish that it is not (i) the command-line itself that fails or (ii) failure to generate the reciprocal comparison command-lines.

These are unit tests, whose coverage may be masked by a passing integration test so removal may not indicate a problem if only the coverage information is being checked. Please can you clarify whether the intent of these skipped tests was maintained elsewhere? If not, do you consider that any of the following hold: (i) the unit test is redundant; (ii) the unit test needs to be updated to reflect the new way that reciprocal comparison commands are generated; (iii) the time taken to update the unit tests would be better spent converting command-line generation to something SLURM-compatible?

If (i) then please delete the tests. If (ii) please update the tests accordingly. If (iii) please ignore - other than to note that (iii) is the case - and work on the new command-line generation.

widdowquinn committed:
update deprecated behaviours in pandas

- the sort_index() call now needs axis as an explicit keyword
- pandas.utils.testing is now pandas.testing

widdowquinn committed:
remove unused import

If an import is no longer required, it should be removed, rather
than commented as "probably not required" or similar.

baileythegreen committed:
Merge branch 'master' of https://github.com/widdowquinn/pyani into anib_123
